add support for the updateNoteFields api

Twinov opened the pull request:
### Description
Adds support for the [updateNoteFields API](https://github.com/FooSoft/anki-connect#updatenotefields) to match the desktop AnkiConnect. There's a couple additional helper functions added as well that were needed to facilitate updateNoteFields.
### Example Request
```js
{
  action: 'updateNoteFields',
  version: 6,
  params: {
    note: {
      id: 1679779279219,
      fields: {
        Source: 'anime',
      },
      picture: [
        {
          data: image as base64 string,
          filename: 'screenshot.jpg',
          fields: ['Source'],
        },
      ],
    },
  },
}
```
### Limitations
Doesn't support URLs for files nor skipHash.
I only tested using images, though I uploaded them as audio/video to test the codepath and it worked fine.

Aquafina-water-bottle commented on app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java:
```diff
@@ -78,6 +78,55 @@ public Long addNote(final Map<String, String> data, String deck_name, String mod
+    public boolean updateNoteFields(long note_id, Map<String, String> new_fields, final Map<String, byte[]> files_to_add, final Map<String, ArrayList<String>> field_to_filenames_img, final Map<String, ArrayList<String>> field_to_filenames_sound) throws Exception {
+        String[] model_field_names = modelAPI.modelFieldNames(noteAPI.getNoteModelId(note_id));
+        String[] card_fields = noteAPI.getNoteFields(note_id);
+        Map<String, String> stored_files = new HashMap<>();
```
It seems like `stored_files` only exists to check for duplicate files. I don't think Anki-Connect handles duplicate files unless specified with an additional [`deleteExisting`](https://github.com/FooSoft/anki-connect/blob/bf5c73b9cdec95337dc356b0d006b7847458c086/plugin/__init__.py#L673) field, so this logic can be safely removed.

Twinov replied:
Actually the map is there because `mediaAPI.storeMediaFile()` doesn't produce the same filename as what's fed into it (at least when I was testing I think), so I cached the filename from the first time `mediaAPI.storeMediaFile()` was called. Otherwise, you'd store the same file multiple times within the same `updateNoteFields()` call under different filenames. More elegant solution would be to find out why `mediaAPI.storeMediaFile()` doesn't store with the exact filename it's called with, but that's why this map exists at least.

Aquafina-water-bottle replied:
Oh interesting, thanks for the clarification! In this case, would it be possible to put a comment explaining exactly that (and when the time comes, add that to the documentation too)?

Twinov replied:
Yeah can do.

Aquafina-water-bottle reviewed (changes requested):
Hey @Twinov, thanks for the PR! Sorry it took a little while, but I left a few comments for the PR.

Additionally, I was talking with @KamWithK yesterday, and due to this PR and #28, we decided that it would be a good idea to provide something like a `docs/api.md`, where people can document what API calls exist, and any edge cases / unhandled parameters / etc. for each api call. Originally, the scope of this project was simply to provide an interface to Yomichan, but as indicated by these PRs. some developers are actually using this now.

There will be a PR soon to add that document, and once that happens, we'd like if you added `updateNoteFields` to that document before getting this properly merged. For example, one edge case that I noticed is that `deleteExisting` isn't handled, so that should be documented.

Nonetheless, this and #28 will be blocked until we get the skeleton of docs/api.md out (I'll aim to make the PR in a day or two).

Twinov commented on app/src/main/java/com/kamwithk/ankiconnectandroid/request_parsers/Parser.java:
```diff
@@ -51,6 +51,77 @@ public static String getNoteQuery(JsonObject raw_data) {
+        return gson.fromJson(raw_data.get("params").getAsJsonObject().get("note").getAsJsonObject().get("fields"), fieldType);
+    }
+
+    public static Map<String, byte[]> getUpdateNoteFilesToAdd(JsonObject raw_data) {
```
What do you think about:
In Parser.java, just have a very dumb method `ArrayList<JsonObject> getUpdateNoteFieldsMedia(JsonObject raw_data)` where the JsonObjects are JsonArrays, and then all the work that's currently done in `getUpdateNoteFilesToAdd()` and `getUpdateNoteMediaFieldsToFilenames()` can be moved into `IntegratedAPI.updateNoteFields()`?

A more sustainable solution IMO would be to make a small class that encodes the audio/video/picture parameter from [here](https://github.com/FooSoft/anki-connect#addnote) and then have the parser parse into that, but tbh I don't have much Java experience and it'd be a lot of work if I'm the only one who's going to be using it, maybe consider it if someone wants to add media support to addNote though.

Aquafina-water-bottle replied:
I think this is fine as long as you mention in a comment that there could be work for improvement. If the storing media files can be moved to the function as mentioned earlier (lines 90-102), then getting the parser to parse the individual components shouldn't be difficult to do anyways.

Twinov replied:
Ended up creating a RequestMedia class for the parser to parse each audio/video/picture into, since it was easier than sending around the metadata to be honest. Let me know what you think.

Twinov commented:
Thanks for the review, your comments are good and I'll commit fixes soon, just want to know if you agree with my parser method suggestion before implementing. 
And yep can add docs once that's in master too.

KamWithK commented:
Just a small reminder @Twinov that the docs pr was recently merged
So once you've made any necessary updates and added docs we can take another look at this PR 😃

Twinov commented:
Made some pretty substantial changes to the updateNoteFields implementation that I think addresses the comments, but definitely needs to be rechecked. I did end up creating a small RequestMedia class to store each media object passed in when parsed.
Will check out the docs format and add soon. (as part of this PR)

Twinov commented on app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java:
```diff
@@ -78,6 +78,55 @@ public Long addNote(final Map<String, String> data, String deck_name, String mod
+            new Handler(Looper.getMainLooper()).post(() -> Toast.makeText(context, "Card updated", Toast.LENGTH_LONG).show());
+        } else {
+            new Handler(Looper.getMainLooper()).post(() -> Toast.makeText(context, "Failed to update card", Toast.LENGTH_LONG).show());
+        }
```
Removed it in the reworked implementation.

Twinov commented on app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java:
```diff
@@ -78,6 +78,55 @@ public Long addNote(final Map<String, String> data, String deck_name, String mod
+                        }
+                        card_fields[i] += "<img src=\"" + stored_name + "\">";
+                    }
+                }
```
New implementation doesn't have the code duplication anymore by changing where the enclosed filename is created, so didn't make a helper function.

Aquafina-water-bottle reviewed (commented):
Appreciate the changes! Everything overall looks fine now, but I do still have a question with how the `stored` variable is used, as commented above.

Twinov commented on app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java:
```diff
@@ -78,6 +79,79 @@ public Long addNote(final Map<String, String> data, String deck_name, String mod
+            if (!media.isStored()) {
+                media.setFilename(mediaAPI.storeMediaFile(media.getFilename(), media.getData()));
+                media.setStored(true);
+            }
```
You're right, everything in RequestMedia is final now and no need to keep metadata about whether file has been stored before.

Twinov commented:
Rebased and resolved conflicts (thus the force push), addressed comments, and added docs. Thanks for the review, good comments.

KamWithK commented on app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/MediaAPI.java:
```diff
@@ -35,6 +35,7 @@ public MediaAPI(Context context) {
      */
     @SuppressLint("SetWorldReadable")
     public String storeMediaFile(String filename, byte[] data) throws IOException {
+        // TODO: investigate why filename gets a number attached to it, i.e. file.png -> file_123456789.png
```
Is this number maybe just a unique identifier?

Twinov replied:
Yes I think it's just a random unique ID, issue is that it's not accurate to how desktop Ankiconnect handles saving files. Desktop version saves files as the passed in filename.

Aquafina-water-bottle commented:
Sorry again for the delay! Just looked through everything again, and it looks good to me! Thanks for responding to all of my earlier nitpicks :tada:
